.github/documents/CONFIG_GUIDE.md
# Config File Guide

<p align="center"><a href="https://github.com/MC-Dashify/launcher/blob/main/.github/documents/CONFIG_GUIDE.md">English</a> · <a href="https://github.com/MC-Dashify/launcher/blob/main/.github/documents/CONFIG_GUIDE.ko_KR.md">한국어</a></p>

## Description by Configuration Settings

The configuration settings file is created in JSON format and is automatically generated with the name `launcher.config.json`.

<table>
<thead>
  <tr>
    <th>Key</th>
    <th>Description</th>
    <th>Default Value</th>
    <th>Note</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td><code>config_version</code></td>
    <td>Indicates the version of the current configuration settings.</td>
    <td><code>1</code></td>
    <td><span style="color:#ff8888">Random changes to the configuration settings version can result in unintended errors.</span></td>
  </tr>
   <tr>
    <td><code>server</code></td>
    <td>Sets the path or link to the server jar file.</td>
    <td><code>https://clip.aroxu.me/download?mc_version=1.19.4</code></td>
    <td><span>Insert the server download link that starts with http(s):// or enter the server file location like following format: file://path/to/server.jar.</span></td>
  </tr>
  <tr>
    <td><code>debug</code></td>
    <td>Set whether to enable JVM debugging.</td>
    <td><code>false</code></td>
    <td><span>N/A</span></td>
  </tr>
  <tr>
    <td><code>debug_port</code></td>
    <td>Sets the JVM debugging port.</td>
    <td><code>5005</code></td>
    <td><span style="color:#ff8888">If the <code>debug</code> option is turned off, this option is ignored.</span></td>
  </tr>
  <tr>
    <td><code>restart</code></td>
    <td>Set whether to restart automatically when the server JVM runtime ends.</td>
    <td><code>true</code></td>
    <td><span style="color:#ff8888">Not applicable for force majeure abnormal termination (power outage, kernel panic, or blue screen).</span></td>
  </tr>
  <tr>
    <td><code>memory</code></td>
    <td>Sets the amount of memory in GB to allocate at server JVM runtime.</td>
    <td><code>2</code></td>
    <td><span style="color:#ff8888">If it does not exceed 2GB, the server may not be running.</span></td>
  </tr>
  <tr>
    <td><code>enable_traffic_monitor</code></td>
    <td>Sets whether the server traffic measurement feature is enabled.</td>
    <td><code>false</code></td>
    <td><span style="color:#ff8888">The feature is still an alpha state. See the README.MD file for a list of known issues.</span></td>
  </tr>
  <tr>
    <td><code>traffic_redirect_port</code></td>
    <td>You must redirect the port to use the Server Traffic Measurement feature. Set the port to use when redirecting.</td>
    <td><code>25555</code></td>
    <td><span style="color:#ff8888">The feature is still an alpha state. See the README.MD file for a list of known issues. If the <code>enable_traffic_monitor</code> option is turned off, this option is ignored.</span></td>
  </tr>
  <tr>
    <td><code>api_port</code></td>
    <td>Sets the port that is exposed to the outside.</td>
    <td><code>8080</code></td>
    <td><span>N/A</span></td>
  </tr>
  <tr>
    <td><code>plugin_api_port</code></td>
    <td>Please make it the same as the port in the Dashify plugin setting that goes into the server. Dashify retrieves information through that port. If the ports do not match, proper information is not delivered.</td>
    <td><code>8081</code></td>
    <td><span>The plug-in settings file is located in the path 'plugins/Dashify/config.yml'.</span></td>
  </tr>
  <tr>
    <td><code>plugins</code></td>
    <td>Please write down the download link of the plug-in to enter the server in array form.</td>
    <td><code>["https://github.com/MC-Dashify/plugin/releases/latest/download/dashify-plugin-all.jar"]</code></td>
    <td><span>If the link is incorrect, it is ignored. The plug-ins listed here are downloaded to the latest version each time launcher is started.</span></td>
  </tr>
  <tr>
    <td><code>jar_args</code></td>
    <td>Set the arguments to run the server.</td>
    <td><code>["nogui"]</code></td>
    <td><span style="color:#ff8888">Invalid arguments are ignored. If you enter some options incorrectly, the server may not function properly.</span></td>
  </tr>
</tbody>
</table>


.github/documents/CONFIG_GUIDE.ko_KR.md
# 구성 설정 파일 가이드

<p align="center"><a href="https://github.com/MC-Dashify/launcher/blob/main/.github/documents/CONFIG_GUIDE.md">English</a> · <a href="https://github.com/MC-Dashify/launcher/blob/main/.github/documents/CONFIG_GUIDE.ko_KR.md">한국어</a></p>

## 구성 설정별 설명

구성 설정 파일은 JSON 형식으로 작성되어 있으며, `launcher.config.json`이라는 이름으로 자동 생성됩니다.

<table>
<thead>
  <tr>
    <th>키</th>
    <th>설명</th>
    <th>기본값</th>
    <th>비고</th>
  </tr>
</thead>
<tbody>
  <tr>
    <td><code>config_version</code></td>
    <td>현재 구성 설정의 버전을 나타냅니다.</td>
    <td><code>1</code></td>
    <td><span style="color:#ff8888">구성 설정 버전을 임의로 변경할 경우 의도치 않은 오류가 일어날 수 있습니다.</span></td>
  </tr>
   <tr>
    <td><code>server</code></td>
    <td>서버 jar 파일의 경로 또는 링크를 설정합니다.</td>
    <td><code>https://clip.aroxu.me/download?mc_version=1.19.4</code></td>
    <td><span>http(s)://로 시작하는 서버 다운로드 링크를 넣거나 file://path/to/server.jar 형식의 서버 파일 위치를 넣습니다.</span></td>
  </tr>
  <tr>
    <td><code>debug</code></td>
    <td>JVM 디버깅 활성화 여부를 설정합니다.</td>
    <td><code>false</code></td>
    <td><span>N/A</span></td>
  </tr>
  <tr>
    <td><code>debug_port</code></td>
    <td>JVM 디버깅 포트를 설정합니다.</td>
    <td><code>5005</code></td>
    <td><span style="color:#ff8888"><code>debug</code> 옵션이 꺼져있을 경우 이 옵션은 무시됩니다.</span></td>
  </tr>
  <tr>
    <td><code>restart</code></td>
    <td>서버 JVM 런타임이 종료되었을 경우 자동으로 재시작 할지 설정합니다.</td>
    <td><code>true</code></td>
    <td><span style="color:#ff8888">불가항적 비정상 종료(정전, 커널 패닉 또는 블루스크린)에 대해서는 해당되지 않습니다.</span></td>
  </tr>
  <tr>
    <td><code>memory</code></td>
    <td>서버 JVM 런타임에 할당할 메모리 양을 GB단위로 설정합니다.</td>
    <td><code>2</code></td>
    <td><span style="color:#ff8888">2GB가 넘지 않을경우 서버가 실행 되지 않을 수 있습니다.</span></td>
  </tr>
  <tr>
    <td><code>enable_traffic_monitor</code></td>
    <td>서버 트래픽 측정 기능을 활성화 할지 설정합니다.</td>
    <td><code>false</code></td>
    <td><span style="color:#ff8888">해당 기능은 아직 alpha 기능입니다. README.MD 파일을 참고하여 알려진 이슈 목록을 확인하세요.</span></td>
  </tr>
  <tr>
    <td><code>traffic_redirect_port</code></td>
    <td>서버 트래픽 측정 기능을 사용하려면 포트를 리다이렉트 해야 합니다. 리다이렉트 할 때 사용할 포트를 설정합니다.</td>
    <td><code>25555</code></td>
    <td><span style="color:#ff8888">해당 기능은 아직 alpha 기능입니다. README.MD 파일을 참고하여 알려진 이슈 목록을 확인하세요. <code>enable_traffic_monitor</code> 옵션이 꺼져있을 경우 이 옵션은 무시됩니다.</span></td>
  </tr>
  <tr>
    <td><code>api_port</code></td>
    <td>밖으로 노출되는 포트를 설정합니다.</td>
    <td><code>8080</code></td>
    <td><span>N/A</span></td>
  </tr>
  <tr>
    <td><code>plugin_api_port</code></td>
    <td>서버에 들어가는 Dashify 플러그인 설정에 들어가있는 포트와 동일하게 해주세요. Dashify는 해당 포트를 통해 정보를 불러옵니다. 포트가 일치하지 않을 경우 제대로 된 정보가 전달되지 않습니다.</td>
    <td><code>8081</code></td>
    <td><span>플러그인 설정 파일은 `plugins/Dashify/config.yml` 경로에 있습니다.</span></td>
  </tr>
  <tr>
    <td><code>plugins</code></td>
    <td>서버에 들어갈 플러그인의 다운로드 링크를 배열 형태로 적어주세요.</td>
    <td><code>["https://github.com/MC-Dashify/plugin/releases/latest/download/dashify-plugin-all.jar"]</code></td>
    <td><span>링크가 올바르지 않을 경우 해당 링크는 무시됩니다. 여기에 적힌 플러그인들은 launcher가 시작될 때 마다 최신 버전으로 다운로드 됩니다.</span></td>
  </tr>
  <tr>
    <td><code>jar_args</code></td>
    <td>서버 실행에 들어갈 arguments를 설정합니다.</td>
    <td><code>["nogui"]</code></td>
    <td><span style="color:#ff8888">유효하지 않은 argument는 무시됩니다. 일부 옵션을 잘못 입력하면 서버가 제대로 작동하지 않을 수 있습니다.</span></td>
  </tr>
</tbody>
</table>


config/config.go
package config

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	"github.com/MC-Dashify/launcher/i18n"
	"github.com/MC-Dashify/launcher/utils"
	"github.com/MC-Dashify/launcher/utils/logger"
	"gopkg.in/yaml.v3"
)

var ConfigContent Config

const configFileName = "launcher.conf.json"
const pluginConfigFileName = "config.yml"

const (
	defaultConfigVersion        = 1
	defaultServer               = "https://clip.aroxu.me/download?mc_version=1.20.1"
	defaultDebug                = false
	defaultDebugPort            = 5005
	defaultEnableTrafficMonitor = true
	defaultRestart              = true
	defaultMemory               = 2
	defaultTrafficRedirectPort  = 25555
	defaultAPIPort              = 8080
	defaultPluginPort           = 8081
)

var (
	defaultPlugins                = []string{"https://github.com/MC-Dashify/plugin/releases/latest/download/dashify-plugin-all.jar"}
	defaultJarArgs                = []string{"nogui"}
	defaultWebConsoleDisabledCmds = []string{"stop"}
)

type Config struct {
	ConfigVersion          int      `json:"config_version"`
	Server                 string   `json:"server"`
	Debug                  bool     `json:"debug"`
	DebugPort              int      `json:"debug_port"`
	Restart                bool     `json:"restart"`
	Memory                 int      `json:"memory"`
	EnableTrafficMonitor   bool     `json:"enable_traffic_monitor"`
	TrafficRedirectPort    int      `json:"traffic_redirect_port"`
	APIPort                int      `json:"api_port"`
	PluginPort             int      `json:"plugin_api_port"`
	Plugins                []string `json:"plugins"`
	JarArgs                []string `json:"jar_args"`
	WebConsoleDisabledCmds []string `json:"webconsole_disabled_cmds"`
}

type PluginConfig struct {
	Key     string `yaml:"key"`
	Enabled bool   `yaml:"enabled"`
}

func LoadConfig() Config {
	var config Config
	defaultConfig := Config{
		ConfigVersion:          defaultConfigVersion,
		Server:                 defaultServer,
		Debug:                  defaultDebug,
		DebugPort:              defaultDebugPort,
		Restart:                defaultRestart,
		Memory:                 defaultMemory,
		TrafficRedirectPort:    defaultTrafficRedirectPort,
		APIPort:                defaultAPIPort,
		PluginPort:             defaultPluginPort,
		Plugins:                defaultPlugins,
		JarArgs:                defaultJarArgs,
		WebConsoleDisabledCmds: defaultWebConsoleDisabledCmds,
	}
	currentPath, _ := os.Getwd()
	configPath := filepath.Join(currentPath, configFileName)

	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		logger.Warn(i18n.Get("config.notfound"))
		saveConfig(defaultConfig)
	}

	configData, loadFileErr := os.ReadFile(configPath)
	if loadFileErr != nil {
		logger.Error(strings.ReplaceAll(i18n.Get("config.loaderror"), "$error", loadFileErr.Error()))
	}

	if strings.TrimSpace(string(configData)) == "" {
		logger.Error(i18n.Get("config.empty"))
		saveConfig(defaultConfig)
		return defaultConfig
	}

	loadConfigErr := json.Unmarshal([]byte(configData), &config)
	if loadConfigErr != nil {
		logger.Fatal(strings.ReplaceAll(i18n.Get("config.invalid"), "$error", loadConfigErr.Error()))
	} else {
		if config.Server == "" {
			logger.Warn(i18n.Get("config.server.empty"))
		}
		if config.TrafficRedirectPort <= 0 || config.TrafficRedirectPort > 65535 {
			logger.Fatal(i18n.Get("config.api_port.invalid"))
		}
		if config.APIPort <= 0 || config.APIPort > 65535 {
			logger.Fatal(i18n.Get("config.api_port.invalid"))
		}
		if config.PluginPort <= 0 || config.PluginPort > 65535 {
			logger.Fatal(i18n.Get("config.plugin_api_port.invalid"))
		}
		if config.DebugPort <= 0 || config.DebugPort > 65535 {
			logger.Fatal(i18n.Get("config.debug_port.invalid"))
		}
		if config.Memory < 2 {
			logger.Fatal(i18n.Get("config.memory.invalid"))
		}
		if config.ConfigVersion != defaultConfigVersion {
			logger.Warn(i18n.Get("config.version.different"))
		}
	}
	return config
}

func saveConfig(data Config) {
	file, _ := json.MarshalIndent(data, "", " ")

	serverConfFile, errGenConf := os.Create(configFileName)

	if errGenConf != nil {
		logger.Fatal(strings.ReplaceAll(i18n.Get("config.create_failed"), "$error", errGenConf.Error()))
	}

	defer serverConfFile.Close()

	_, errWrtConf := serverConfFile.Write(file)

	if errWrtConf != nil {
		logger.Fatal(strings.ReplaceAll(i18n.Get("config.write_failed"), "$error", errWrtConf.Error()))
	}
	logger.Info(i18n.Get("config.created"))
}

func GetPluginConfig() PluginConfig {
	var pluginConfig PluginConfig
	currentPath, _ := os.Getwd()
	configPath := filepath.Join(currentPath, "plugins", "Dashify", pluginConfigFileName)

	if _, err := os.Stat(configPath); os.IsNotExist(err) {
		randomString := utils.GenerateRandomString(64)
		hashedString := utils.GenerateBCryptString(randomString)
		pluginConfig = PluginConfig{
			Key: hashedString,
		}
		savePluginConfig(pluginConfig)
	}

	configData, loadFileErr := os.ReadFile(configPath)
	if loadFileErr != nil {
		logger.Error(loadFileErr.Error())
	}

	if strings.TrimSpace(string(configData)) == "" {
		randomString := utils.GenerateRandomString(64)
		hashedString := utils.GenerateBCryptString(randomString)
		pluginConfig = PluginConfig{
			Key: hashedString,
		}
		savePluginConfig(pluginConfig)
	}

	loadConfigErr := yaml.Unmarshal([]byte(configData), &pluginConfig)
	if loadConfigErr != nil {
		logger.Fatal(strings.ReplaceAll(i18n.Get("config.invalid"), "$error", loadConfigErr.Error()))
	} else {
		if pluginConfig.Key == "" {
			randomString := utils.GenerateRandomString(64)
			hashedString := utils.GenerateBCryptString(randomString)
			pluginConfig = PluginConfig{
				Key: hashedString,
			}
			savePluginConfig(pluginConfig)
		}
	}
	return pluginConfig
}

func savePluginConfig(data PluginConfig) {
	currentPath, _ := os.Getwd()
	configPath := filepath.Join(currentPath, "plugins", "Dashify", pluginConfigFileName)

	utils.CheckIsExist(filepath.Join(currentPath, "plugins", "Dashify"))

	file, _ := yaml.Marshal(data)

	serverConfFile, errGenConf := os.Create(configPath)

	if errGenConf != nil {
		logger.Fatal(errGenConf.Error())
	}

	defer serverConfFile.Close()

	_, errWrtConf := serverConfFile.Write(file)

	if errWrtConf != nil {
		logger.Fatal(errWrtConf.Error())
	}
}
